### ISSUE TITLE
ArrowApply? Monadic instance of MSF?

### ISSUE THREAD

**Issue by freckletonj** [ivanperez-keera/dunai#237]
To get acquainted with `dunai`, I'm trying to write a stream function that returns/consumes stream functions. 

```haskell
x :: SF a (SF b Int)
x = arr $ const (arr $ const 42)

-- or

y :: SF (SF a b) b
y = undefined
```

I think this would require `ArrowApply`, or a monadic `join`.

I've tried writing a Monadic instance but failed.

I'm guessing this is impossible for a reason though? If so, what is that reason, or is it documented anywhere?

**Comment by turion** [comment]
I think the general wisdom is that `ArrowApply` and `Monad` are in principle possible, but inefficient (it gets slower the longer the MSF runs) and not useful, and it's not exactly what you'd expect.

As a warmup, you can try to implement `Monad` for streams:

```
data Stream a = Stream { head :: a, tail :: Stream a }
```
The idea is to use the "diagonal":

```
instance Functor Stream where
  fmap f Stream { .. } = Stream (f head) $ f <$> tail

instance Monad Stream where
  return a = go where go = Stream a go
  join (Stream stream streams) = Stream (head stream) $ join $ tail <$> streams
```

That's the rough idea, at least. As the next step, you can implement the `Monad` instance for `StreamT`:

```
data StreamT m a = StreamT { runStreamT :: m (StreamT' m a) }
data StreamT' m a = StreamT' { head :: a, tail :: StreamT m a }
```

Then one can realise that `MSF m a b` is isomorphic to `StreamT (ReaderT a m) b`, which gives you the `Monad` instance. In particular, `SF a b` is isomorphic to `StreamT (Reader a) b`. But note that it's _not_ the same instance as you'd get if you had `ArrowApply`, because `MSF m a b` is _not_ a Kleisli arrow, i.e. it's not of the form `a -> t m b` for some monad transformer `t`. The reason is that an `MSF` will receive many inputs, whereas a Kleisli arrow only ever receives one.

So, in short: `Monad` yes, but awkward and not useful. `ArrowApply` probably no, and different from `Monad`.
I don't know whether this is documented anywhere properly, because it's kind of a negative result.
For some time I've wanted to write an article about applicative transformers, because `SF` is one.

**Comment by freckletonj** [comment]
thanks @turion , I'll keep working with this :D

**Comment by freckletonj** [comment]
If anyone else stumbles upon this question, I'm just reading [Generalizing Monads to Arrows](http://www.cse.chalmers.se/~rjmh/Papers/arrows.pdf) and it explains a bit more about `ArrowApply` and `Monad` versions of arrows.
